### ISSUE TITLE
Configuring a global custom ITypeConverter in AutoMapper 10

### ISSUE THREAD

**Issue by felipeoriani** [AutoMapper/AutoMapper#3557]
Hi guys,

I am trying to upgrade the AutoMapper version into our project here. We were using `AutoMapper 5.x` with `.Net Framework` and we are working on a new version with `AutoMapper 10`. Since the static mapper was removed from the AutoMapper, we have defined our mappings with subclasses of `Profile` class and configure one for each entity/command. Then we inject an instance of `IMapper` and it works.

In the past, [Jimmy help](https://github.com/AutoMapper/AutoMapper/issues/1871#issuecomment-269524441) me with a custom conversion from `long` to `Entity`, because we are going to take an ID from the front-end app and we need to convert it with NHibernate into our Entity. I want to define a custom map for each entity based on ID as a source. I want to be able to configure our map something like this:

```
CreateMap<WorkOrderCommand, WorkOrderEntity>()
    .ForMember(u => u.City, opt => opt.MapFrom(c => c.CityId))
    .ForMember(u => u.Employee, opt => opt.MapFrom(c => c.EmployeeId));
```

Where `City` and `Employee` are entities and `CityId` and `EmployeeId` are Ids represented by `long`. We will have more then 1000 mappings configurations like this over the entire project. I believe `ITypeConverter` I can do it but I do not know how to configure it on the new AutoMapper versions.

I prepared a gist of what I am trying based on my experience (as called before creating this thread).
https://gist.github.com/felipeoriani/248bdfe945f76810fbbd2372ba415a62

**The problem** is when the runtime hits the point where we call the `IMapper.Map` method, it does not convert the Id and give me an exception saying that there is no configuration for long to that entity. How can I get this configuration globally working?

Thank you.

**Comment by jbogard** [comment]
You need to configure your type converters. AutoMapper can't close "partially" closed generic types like you have above automatically, so you'll need to register each entity as its own type converter. Something like

```csharp
cfg.CreateMap<long, Customer>().ConvertUsing<EntityConverter<long, Customer>>();
cfg.CreateMap<long, Order>().ConvertUsing<EntityConverter<long, Order>>();
// etc
```

**Comment by felipeoriani** [comment]
Hi Jimmy, thanks for the anwser! I defined as you said and it works. The problem to me on defining it using the strong generic types is the amount of entities (more than 100 and counting...).

I got it working but I am not sure if it is ok with autoMapper. I was having an issue with the `ServiceCollection` to inject the dependency on the `EntityConverter` instance. I am  injecting the an abstract that encapsulate the `ISession` on the Converter. Now I am able to use the non-generic versions.

Thank you!
